qiime2/sdk/plugin_manager.py

```diff
@@ -142,34 +144,75 @@ def get_semantic_types(self):
 
         return types
 
+    class FormatFilters(enum.Flag):
+        EXPORTABLE = enum.auto()
+        IMPORTABLE = enum.auto()
+
     # TODO: Should plugin loading be transactional? i.e. if there's
     # something wrong, the entire plugin fails to load any piece, like a
     # databases rollback/commit
 
-    def get_formats(self, *, include_all=False, importable=False,
-                    exportable=False, canonical_format=False):
-        if include_all is True and canonical_format is True or include_all \
-                is True and importable is True or include_all is True and \
-                exportable is True:
-            raise ValueError("If all formats are requested, other formats "
-                             "cannot be included as a result.")
-
-        elif include_all is True:
-            return self.formats
+    def get_formats(self, *, filter=None, semantic_type=None):
 
         result_formats = set()
 
-        if importable is True:
-            result_formats = result_formats.union(self._importable)
+        if semantic_type is None:
+            if filter is None:
+                return self.formats
+
+            if self.FormatFilters.IMPORTABLE in filter:
+                result_formats = result_formats.union(self._importable)
+
+            if self.FormatFilters.EXPORTABLE in filter:
+                result_formats = result_formats.union(self._exportable)
+
+        else:
+
+            if semantic_type not in self.get_semantic_types():
+                raise ValueError("The semantic type provided: %s is not "
+                                 "valid.", (semantic_type))
+
+            if filter is None:
+                filter = self.FormatFilters.IMPORTABLE | \
+                            self.FormatFilters.EXPORTABLE
+
+                for type_format in self.type_formats:
+                    if semantic_type in type_format.type_expression:
+                        result_formats = result_formats.union(
+                            type_format.format)
+
+                    if self.FormatFilters.IMPORTABLE in filter:
 
-        if exportable is True:
-            result_formats = result_formats.union(self._exportable)
+                        result_formats = result_formats.union(
+                                self.get_format_to(type_format.format))
 
-        if canonical_format is True:
-            result_formats = result_formats.union(self._canonical_formats)
+                    if self.FormatFilters.EXPORTABLE in filter:
+
+                        result_formats = result_formats.union(
+                            self.get_format_from(type_format.format))
 
         return result_formats
 
+    def get_format_from(self, canonical_format):
+
+        formats = set()
```

Comment:
Stylistic, but I just wanted to point out that a set comprehension could work here, and is generally seen as a more "pythonic" or idiomatic approach (but again, that is really a minor "style" point).